Automation menu - base PR (#2593)
astronomer/docs: astro/automation-authentication.md

---

> sidebar_label: 'Authenticate an automated workflow'

Reviewer: TBH -- This doc is a little confusing to me in that it's both:

- "What are API keys" (duplicate of API tokens docs)
- "How do I generally automate things on Astro" (duplicate of automation overview?)

Author: This doc is "how do you set up your automation environment to connect to Astro"

title: 'Authenticate to Astro for automated workflows'
id: automation-authentication
description: Learn about all possible ways that you can authenticate to Astro from the Astro CLI and automation tools.
---

> Before you automate your workflows on Astro, you must prove to Astro that your automation tool is using the right identity and access to interact with specific Astro resources. You can do this using the Astro CLI and API tokens. You must complete the following actions to authenticate to Astro:

Reviewer: Also, I'd like to see a: "This guide helps you...". Just since the scope of the doc is a bit unclear, and there's obviously lots of links out to other places and related content that a user might get confused about

> Before you automate your workflows on Astro, you must prove to Astro that your automation tool is using the right identity and access to interact with specific Astro resources. You can do this using the Astro CLI and API tokens. You must complete the following actions to authenticate to Astro:

Author: I think some of the expository info you want to cover here is why the "Overview" doc exists. If someone lands on this doc and doesn't know what they would be automating, a common action would be to go to the Overview to get more context. I don't want to overload the top of this doc with context that's explained elsewhere.

> - Create a credential in Astro.

Reviewer: Doesn't have to be this exact wording, but some "so what" thing if this is a conceptual doc

> - Create a credential in Astro.

Author: This is not a conceptual doc - it's explicit purpose is to give steps for authenticating a tool to Astro, so that would make it a process doc

- Make the credentials accessible to the Astro CLI installed in your automation environment.
- Install the Astro CLI in an environment from where you want to run automated processes, such as a GitHub Actions.

> Astro's authentication process is based on [Auth0 Identifier First Authentication flow](https://auth0.com/docs/authenticate/login/auth0-universal-login/identifier-first). This process doesn't provide authorization and doesn't care what a user can do in Astro. To manage authorization in Astro, see [User permissions](user-permissions.md).

Author: Manmeet added this paragraph because it's apparently a common question that comes up amongst customers and the field. It's meant to both

- Say the doc is NOT about authorization.
- Link users to the right doc for authorization management (user permissions).

> ## Create a credential

Reviewer: ```suggestion
## Create an API token or API key
```

Let's just call it this explicitly instead of introducing a third term?

You can use any of the following credentials to authenticate in an automated process:

- A Deployment API key. See [Create a Deployment API key](api-keys.md).
- A Workspace API token. See [Create a Workspace API token](workspace-api-tokens.md).
- An Organization API token. See [Create an Organization API token](organization-api-tokens.md).

> You can create API tokens with role-based access permissions, ensuring that they are not over-provisioned for their intended use. When you chose a credential, you should grant it the least amount of permissions possible for the action it's used for. For example, instead of creating an Organization API token to automate actions in two separate Workspaces, you can create two separate Workspace API tokens for each Workspace.

Reviewer: Agreed, "over-provisioned" is a bit weird here. Can we be more specific, clear here?

## Make credentials accessible to your automation environment

> The Astro CLI uses reserved environment variables to read your credential information to authenticate to Astro. You must set these environment variables in your automation environment. 

Reviewer: "Reserved"? Might be good to start this sentence with env-vars first language vs. "this is how the Astro CLI works"

What's the purpose of Env Vars here? Also, we now keep using "in your automation environment" and I don't super know what that means. Let's give examples like, "GitHub secrets"?

Author: Tried to make the environment variable workflow more clear

We include the example in the top: `automation environment, such as a GitHub Actions.`

:::caution

Because these environment variables store sensitive credentials, Astronomer recommends encrypting the variable values before using them in your script. You can do this either natively in your automation tool or in a secrets backend. 

:::

To use a Deployment API key as an authentication credential, set the following environment variables in your script:

```bash
ASTRONOMER_KEY_ID=<your-api-key-id>
ASTRONOMER_KEY_SECRET=<your-api-key-secret>
```

To use a Workspace or Organization API token as an authentication credential, set the following environment variable in your script: 

```bash
ASTRO_API_TOKEN=<your-api-token>
```

## Install the Astro CLI in your automation tool

To manage your Astro workflows programmatically, you must install the Astro CLI in the environment which will run the workflows. Typically, this requires running `curl -sSL install.astronomer.io | sudo bash -s` or an equivalent installation command before your process starts. See [CI/CD templates](ci-cd-templates/template-overview.md) for examples of how to install the Astro CLI in different version management and workflow automation environments. 

> 

Reviewer: Now what? No other "next steps" or links?

---

> sidebar_label: 'Authentication'

Author: ```suggestion
sidebar_label: 'Authenticate an automated workflow'
```

> sidebar_label: 'Authenticate an automated workflow'

Author: ```suggestion
sidebar_label: 'Authenticate an automation environment'
```
I still don't like the current title, as you are technically authenticating the tool, not the script. But I am open to other suggestions here.

title: 'Authenticate to Astro for automated workflows'
id: automation-authentication
description: Learn about all possible ways that you can authenticate to Astro from the Astro CLI and automation tools.
---

> Astro's authentication process is based on [Auth0 Identifier First Authentication flow](https://auth0.com/docs/authenticate/login/auth0-universal-login/identifier-first). This process doesn't provide authorization and doesn't care what a user can do in Astro. To manage authorization in Astro, see [User permissions](./user-permissions.md).

Author: ```suggestion
```

Before you automate your workflows on Astro, you must prove to Astro that your automation tool is using the right identity and access to interact with specific Astro resources. You can do this using the Astro CLI and API tokens. You must complete the following actions to authenticate to Astro:

- Create a credential in Astro.
- Make the credentials accessible to the Astro CLI installed in your automation environment.
- Install the Astro CLI in an environment from where you want to run automated processes, such as a GitHub Actions.

> 

Author: ```suggestion

Astro's authentication process is based on [Auth0 Identifier First Authentication flow](https://auth0.com/docs/authenticate/login/auth0-universal-login/identifier-first). This process doesn't provide authorization and doesn't care what a user can do in Astro. To manage authorization in Astro, see [User permissions](user-permissions.md).

```

:::

## Make credentials accessible to your automation environment

The Astro CLI uses reserved environment variables to read your credential information to authenticate to Astro. You must set these environment variables in your automation environment. 

:::caution

Because these environment variables store sensitive credentials, Astronomer recommends encrypting the variable values before using them in your script. You can do this either natively in your automation tool or in a secrets backend. 

:::

To use a Deployment API key as an authentication credential, set the following environment variables in your script:

```bash
ASTRONOMER_KEY_ID=<your-api-key-id>
ASTRONOMER_KEY_SECRET=<your-api-key-secret>
```

To use a Workspace or Organization API token as an authentication credential, set the following environment variable in your script: 

```bash
ASTRO_API_TOKEN=<your-api-token>
```

## Install the Astro CLI in your automation tool

> To manage your Astro processes programmatically, you must install the Astro CLI in the environment which will execute the actions. Typically, this requires running `curl -sSL install.astronomer.io | sudo bash -s` or an equivalent installation command before your process starts. See [CI/CD templates](ci-cd-templates/template-overview.md) for examples of how to install the Astro CLI in different version management and workflow automation environments. 

Author: ```suggestion
To manage your Astro workflows programmatically, you must install the Astro CLI in the environment which will run the workflows. Typically, this requires running `curl -sSL install.astronomer.io | sudo bash -s` or an equivalent installation command before your process starts. See [CI/CD templates](ci-cd-templates/template-overview.md) for examples of how to install the Astro CLI in different version management and workflow automation environments. 
```

---
sidebar_label: 'Authenticate an automated workflow'

> title: 'Authenticate to Astro for automated workflows'

Author: It's not just CI/CD tho - scripts can cover a wide variety of use cases beyond CI/CD

- Install the Astro CLI in your automation environment, such as a GitHub Actions.
- Make the key or token accessible to the Astro CLI installed in your automation environment.

After you complete this setup, your automation environment is authenticated to Astro. You can then write and run scripts to manage your Deployments through CI/CD.

Astro's authentication process is based on the [Auth0 Identifier First Authentication flow](https://auth0.com/docs/authenticate/login/auth0-universal-login/identifier-first). This process doesn't provide authorization and isn't affected by what a user can do in Astro. To manage authorization in Astro, see [User permissions](user-permissions.md).

## Step 1: Create an API token or API key

You can use any of the following credentials to authenticate in an automated process:

- A Deployment API key. See [Create a Deployment API key](api-keys.md).
- A Workspace API token. See [Create a Workspace API token](workspace-api-tokens.md).
- An Organization API token. See [Create an Organization API token](organization-api-tokens.md).

When you create an API token or key for your environment, keep the following best practices in mind:

> - Always give your API token or key the minimum permissions required to perform an action. This improves control and security over your Astro components. For example, instead of creating an Organization API token to automate actions in two separate Workspaces, create two separate Workspace API tokens for each Workspace.

Reviewer: ```suggestion
- Always give your API token or key the minimum permissions required to perform an action. This improves control and security over your Astro components. For example, instead of creating an Organization API token to automate actions in two Workspaces, create two separate Workspace API tokens for each Workspace.
```
